File: cpp/BinaryPow.cpp
#include <iostream>

using namespace std;

int pow(int x, int n, int MOD) {
	long long y = x;
	int res = 1;
	for (; n > 0; n >>= 1) {
		if (n & 1)
			res = res * y % MOD;
		y = y * y % MOD;
	}
	return res;
}

int main() {
	const int MOD = 1000000007;
	int x = pow(2, 10, MOD);
	cout << x << endl;
}

File: java/src/BinaryPow.java
public class BinaryPow {

	public static int pow(int x, int n, int mod) {
		int res = 1;
		for (long p = x; n > 0; n >>= 1, p = (p * p) % mod) {
			if ((n & 1) != 0) {
				res = (int) (res * p % mod);
			}
		}
		return res;
	}

	// usage example
	public static void main(String[] args) {
		System.out.println(8 == pow(2, 3, 100));
	}
}
